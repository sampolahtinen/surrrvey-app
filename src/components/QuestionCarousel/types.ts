import { APISurveyQuestion } from "../../api/types";

export interface QuestionCarouselProps {
  questions: APISurveyQuestion[];
  onOptionSelect: (param: string) => void;
  selectedOption: string;
  onQuestionChange: any;
}
