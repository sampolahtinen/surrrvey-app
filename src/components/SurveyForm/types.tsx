import { APISurveyDetailsResponse } from "../../api/types";

export interface SurveyFormProps {
  title: string;
  subtitle: string;
  questions: APISurveyDetailsResponse["questions"];
  onSubmit: (params: any) => any;
  isCompleted: boolean;
  errorMessage?: string | boolean;
}
