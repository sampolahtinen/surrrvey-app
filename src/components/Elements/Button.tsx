import styled from "styled-components";
import { colors } from "../../styles/colors";

export const Button = styled.button`
  width: 100px;
  height: 40px;
  font-size: 16px;
  border-radius: 12px;
  border: 1px ${colors.lila} solid;
  background-color: ${colors.lila};
  color: ${colors.offWhite};
  cursor: pointer;
`;
