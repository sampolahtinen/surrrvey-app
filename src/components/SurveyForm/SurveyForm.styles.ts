import styled from "styled-components";
import { animated } from "react-spring";
import { colors } from "../../styles/colors";

export const SuccessMessage = styled(animated.div)`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: ${colors.white};
  flex-wrap: wrap;
  span {
    display: block;
    color: ${colors.green};
    font-size: 64px;
    font-weight: 600;
    width: 100%;
    text-align: center;
  }
  a {
    font-size: 20px;
  }
`;

export const Wrapper = styled.div`
  position: relative;
  width: 500px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 2rem;
  z-index: 10;
`;

export const Title = styled.h1`
  margin-bottom: 2.4rem;
`;

export const Subtitle = styled.h2`
  margin-bottom: 6.4rem;
  color: ${colors.grey};
  font-weight: 300;
  opacity: 0.5;
`;
