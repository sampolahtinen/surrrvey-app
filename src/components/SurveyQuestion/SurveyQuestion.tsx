import React from "react";
import {
  QuestionsContainer,
  QuestionTitle,
  QuestionList,
  Option
} from "./SurveyQuestion.styles";
import { SurveyQuestionProps } from "./types";

const SurveyQuestion: React.FC<SurveyQuestionProps> = ({
  options,
  selectedOption,
  title,
  onOptionSelect
}) => {
  return (
    <QuestionsContainer style={{ marginBottom: "3.2rem" }}>
      <QuestionTitle>{title}</QuestionTitle>
      <QuestionList>
        {options.map((option: string) => (
          <Option
            key={title + option}
            onClick={() => onOptionSelect(option)}
            selected={selectedOption === option}
          >
            {option}
          </Option>
        ))}
      </QuestionList>
    </QuestionsContainer>
  );
};

export default SurveyQuestion;
