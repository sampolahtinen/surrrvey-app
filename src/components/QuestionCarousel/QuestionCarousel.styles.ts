import styled, { css } from "styled-components";

import { Button } from "../Elements/Button";

export const ButtonsWrapper = styled.div`
  display: flex;
  justify-content: space-between;
`;

export const NextQuestionButton = styled(Button)`
  ${props =>
    props.disabled &&
    css`
      background-color: grey;
      border: 0;
    `}
`;

export const QuestionCarouselContainer = styled.div`
  width: 100%;
`;

export const PrevQuestionButton = styled(NextQuestionButton)``;
