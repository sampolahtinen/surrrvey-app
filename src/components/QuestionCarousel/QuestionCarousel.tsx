import React, { FC, useState, useEffect } from "react";
import SurveyQuestion from "../SurveyQuestion";
import styled, { css } from "styled-components";
import { Button } from "../Elements/Button";
import { QuestionCarouselProps } from "./types";

export const ButtonsWrapper = styled.div`
  display: flex;
  justify-content: space-between;
`;

export const NextQuestionButton = styled(Button)`
  ${props =>
    props.disabled &&
    css`
      background-color: grey;
      border: 0;
    `}
`;

export const PrevQuestionButton = styled(NextQuestionButton)``;

const QuestionCarousel: FC<QuestionCarouselProps> = ({
  questions,
  onOptionSelect,
  selectedOption,
  onQuestionChange
}) => {
  const [currentQuestionIndex, setCurrentQuestion] = useState(0);

  useEffect(() => {
    onQuestionChange(currentQuestionIndex);
  }, [currentQuestionIndex, onQuestionChange]);

  const handleNext = () => setCurrentQuestion(currentQuestionIndex + 1);
  const handlePrev = () => setCurrentQuestion(currentQuestionIndex - 1);

  return (
    <div>
      <SurveyQuestion
        title={questions[currentQuestionIndex].title}
        options={questions[currentQuestionIndex].options}
        selectedOption={selectedOption}
        onOptionSelect={onOptionSelect}
      />
      <ButtonsWrapper>
        <PrevQuestionButton
          onClick={handlePrev}
          disabled={currentQuestionIndex === 0}
        >
          Back
        </PrevQuestionButton>
        <NextQuestionButton
          onClick={handleNext}
          disabled={currentQuestionIndex === questions.length - 1}
        >
          Next
        </NextQuestionButton>
      </ButtonsWrapper>
    </div>
  );
};

export default QuestionCarousel;
