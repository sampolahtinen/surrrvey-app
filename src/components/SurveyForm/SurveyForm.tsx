import React, { FC, useState } from "react";
import { SurveyFormProps } from "./types";
import { Link } from "react-router-dom";
import { config, useTransition } from "react-spring";
import QuestionCarousel from "../QuestionCarousel";
import { Subtitle, Form, Title, SuccessMessage } from "./SurveyForm.styles";
import { Button } from "../Elements/Button";
import { APISurveyCompletion } from "../../api/types";
import { colors } from "../../styles/colors";

const SurveyForm: FC<SurveyFormProps> = ({
  title,
  subtitle,
  questions,
  onSubmit,
  isCompleted,
  errorMessage
}) => {
  const [currentQuestion, setCurrentQuestion] = useState(0);

  const initialState = {
    completion: []
  };

  const [selectedOptions, setSelectedOptions] = useState<APISurveyCompletion>(
    initialState
  );

  const getQuestionAnswer = (questionId: string) =>
    selectedOptions.completion.find(elem => elem.question_id === questionId)
      ?.value;

  const getQuestionIndex = (questionId: string) => {
    const answers = selectedOptions.completion;
    const questionIndex = answers.findIndex(
      elem => elem.question_id === questionId
    );
    return questionIndex;
  };

  const handleOptionSelection = (question: string) => {
    const currentQuestionId = questions[currentQuestion].id;
    const questionIndex = getQuestionIndex(currentQuestionId);
    // If question has not been answered, append it to the completion array
    if (questionIndex === -1) {
      setSelectedOptions({
        completion: [
          ...selectedOptions.completion,
          {
            question_id: currentQuestionId,
            value: question
          }
        ]
      });
      // if question is found from completion, update its value
    } else {
      const nextState = selectedOptions.completion;
      nextState[questionIndex] = {
        question_id: currentQuestionId,
        value: question
      };
      setSelectedOptions({
        completion: nextState
      });
    }
  };

  const handleSubmit = (e: any) => {
    e.preventDefault();
    onSubmit(selectedOptions);
  };

  const transitions = useTransition(isCompleted, null, {
    from: { opacity: 0, transform: "scaleY(0)", transformOrigin: "top" },
    enter: { opacity: 1, transform: "scaleY(1)", transformOrigin: "top" },
    leave: { opacity: 0, transform: "scaleY(0)", transformOrigin: "bottom" },
    config: {
      ...config.stiff,
      clamp: true
    }
  });

  return (
    <Form>
      <Title>{title}</Title>
      <Subtitle>{subtitle}</Subtitle>
      <QuestionCarousel
        questions={questions}
        onOptionSelect={handleOptionSelection}
        selectedOption={getQuestionAnswer(questions[currentQuestion].id) || ""}
        onQuestionChange={(nextQuestionIndex: number) => {
          setCurrentQuestion(nextQuestionIndex);
        }}
      />
      {currentQuestion === questions.length - 1 && (
        <Button
          type="submit"
          onClick={handleSubmit}
          style={{ marginTop: "2rem", width: "150px", height: "50px" }}
        >
          Submit
        </Button>
      )}
      {transitions.map(
        ({ item, props }) =>
          item && (
            <SuccessMessage style={props}>
              <span>Success!</span>
              <span style={{ fontSize: "20px", color: colors.darkGrey }}>
                Your response was delivered.
              </span>
              <Link to="/surveys">Go back to Survey list</Link>
            </SuccessMessage>
          )
      )}
      {errorMessage && <span>{errorMessage}</span>}
    </Form>
  );
};

export default SurveyForm;
