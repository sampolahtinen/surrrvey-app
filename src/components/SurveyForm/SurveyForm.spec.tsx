import React from "react";
import { mount } from "enzyme";
import SurveyForm from "./SurveyForm";
import { Title, Subtitle, Form } from "./SurveyForm.styles";
import { Button } from "../Elements/Button";
import { Option } from "../SurveyQuestion/SurveyQuestion.styles";
import { NextQuestionButton } from "../QuestionCarousel/QuestionCarousel.styles";

const mockQuestions = [
  {
    id: "001",
    title: "firstQuestion",
    options: ["test1", "test2"]
  },
  {
    id: "002",
    title: "secondQuestion",
    options: ["test1", "test2"]
  }
];

describe("SurveyForm  Component", () => {
  const mockSubmitFunc = jest.fn();
  const wrapper = mount(
    <SurveyForm
      title="testTitle"
      subtitle="testSubtitle"
      isCompleted={false}
      questions={mockQuestions}
      onSubmit={mockSubmitFunc}
    />
  );

  it("renders", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("displays title", () => {
    expect(wrapper.find(Title).text()).toEqual("testTitle");
  });

  it("displays subtitle", () => {
    expect(wrapper.find(Subtitle).text()).toEqual("testSubtitle");
  });

  it("selects a question option", () => {
    // Find a option
    const firstOption = wrapper.find(Option).first();
    // Click it
    firstOption.simulate("click");
    // Find the first option again and make assertion
    expect(
      wrapper
        .find(Option)
        .first()
        .prop("selected")
    ).toBeTruthy();
  });

  it("changes selected option", () => {
    const secondOption = wrapper.find(Option).last();
    secondOption.simulate("click");

    // Assert that previously selected option gets unselected
    expect(
      wrapper
        .find(Option)
        .first()
        .prop("selected")
    ).toBeFalsy();

    // Assert that the second option gets selected
    expect(
      wrapper
        .find(Option)
        .at(1)
        .prop("selected")
    ).toBeTruthy();
  });

  it("displays submit button when last question ", () => {
    // Find carousel next button and click it
    wrapper.find(NextQuestionButton).simulate("click");
    // When last element of the carousel is in the viewport, display submit button
    expect(wrapper.find(Button).exists()).toBeTruthy();
    // Click submit
    wrapper.find(Button).simulate("click");
    // Should trigger onSubmit prop
    expect(mockSubmitFunc).toHaveBeenCalled();
    // console.log(mockSubmitFunc.mock.calls[0][0]);
  });
});
