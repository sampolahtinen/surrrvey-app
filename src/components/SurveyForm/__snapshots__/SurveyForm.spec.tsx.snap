// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SurveyForm  Component renders 1`] = `
<SurveyForm
  isCompleted={false}
  onSubmit={[MockFunction]}
  questions={
    Array [
      Object {
        "id": "001",
        "options": Array [
          "test1",
          "test2",
        ],
        "title": "firstQuestion",
      },
      Object {
        "id": "002",
        "options": Array [
          "test1",
          "test2",
        ],
        "title": "secondQuestion",
      },
    ]
  }
  subtitle="testSubtitle"
  title="testTitle"
>
  <styled.form>
    <form
      className="sc-fzXfLO fWeOZL"
    >
      <styled.h1>
        <h1
          className="sc-fzXfLP xGwGK"
        >
          testTitle
        </h1>
      </styled.h1>
      <styled.h2>
        <h2
          className="sc-fzXfLQ jMdyVK"
        >
          testSubtitle
        </h2>
      </styled.h2>
      <QuestionCarousel
        onOptionSelect={[Function]}
        onQuestionChange={[Function]}
        questions={
          Array [
            Object {
              "id": "001",
              "options": Array [
                "test1",
                "test2",
              ],
              "title": "firstQuestion",
            },
            Object {
              "id": "002",
              "options": Array [
                "test1",
                "test2",
              ],
              "title": "secondQuestion",
            },
          ]
        }
        selectedOption=""
      >
        <div>
          <SurveyQuestion
            onOptionSelect={[Function]}
            options={
              Array [
                "test1",
                "test2",
              ]
            }
            selectedOption=""
            title="firstQuestion"
          >
            <styled.div
              style={
                Object {
                  "marginBottom": "3.2rem",
                }
              }
            >
              <div
                className="sc-AykKE FJqhk"
                style={
                  Object {
                    "marginBottom": "3.2rem",
                  }
                }
              >
                <styled.h3>
                  <h3
                    className="sc-AykKD bLtnPA"
                  >
                    firstQuestion
                  </h3>
                </styled.h3>
                <styled.ul>
                  <ul
                    className="sc-AykKC fmcvyS"
                  >
                    <styled.li
                      key="firstQuestiontest1"
                      onClick={[Function]}
                      selected={false}
                    >
                      <li
                        className="sc-AykKF iPtPKR"
                        onClick={[Function]}
                        selected={false}
                      >
                        test1
                      </li>
                    </styled.li>
                    <styled.li
                      key="firstQuestiontest2"
                      onClick={[Function]}
                      selected={false}
                    >
                      <li
                        className="sc-AykKF iPtPKR"
                        onClick={[Function]}
                        selected={false}
                      >
                        test2
                      </li>
                    </styled.li>
                  </ul>
                </styled.ul>
              </div>
            </styled.div>
          </SurveyQuestion>
          <styled.div>
            <div
              className="sc-AykKH iijtjJ"
            >
              <Styled(Styled(styled.button))
                disabled={true}
                onClick={[Function]}
              >
                <button
                  className="sc-AykKG sc-AykKI sc-AykKJ bxMDJn"
                  disabled={true}
                  onClick={[Function]}
                >
                  Back
                </button>
              </Styled(Styled(styled.button))>
              <Styled(styled.button)
                disabled={false}
                onClick={[Function]}
              >
                <button
                  className="sc-AykKG sc-AykKI kANYNs"
                  disabled={false}
                  onClick={[Function]}
                >
                  Next
                </button>
              </Styled(styled.button)>
            </div>
          </styled.div>
        </div>
      </QuestionCarousel>
    </form>
  </styled.form>
</SurveyForm>
`;
